### PR TITLE
MindReading PartII

### PR DESCRIPTION
**Pull request by cyl19970726**

### CODE REVIEW

**Comment by qizhou** [comment]
I would suggest adding a brief summary of what the PR is doing.

File: core/state_processor.go [hunk 1]
```diff
@@ -121,7 +121,6 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 				}
 			} else {
 				// a non-validator fails to replay the MR data
-				// TODO: make sure all mrOutput's are consumed?
```
Comment (qizhou) [inline core/state_processor.go]: I don't think we have address the TODO, right?

**Comment by cyl19970726** [comment]
This PR mainly completes the following functions:
1. Add MindReading to the proposal block process of the proposer-validator-node （core/state_process.go miner/worker.go）
2. Add MindReading to the verification block process of the non-proposer-validator-node (core/state_process.go)
3. Add MindReading to the block synchronization process of the non-validator node(core/state_process.go)
4. Modify tracerApi to support MindReading(eth/)
5. Optimize crossChainCall precompile contract (core/vm/contracts.go)

File: core/state_processor.go (hunk 1 above)
Comment (cyl19970726) [reply core/state_processor.go]: I think the following code can make sure all the mrOuput is consumed.
core/state_process.go line-127
```
	// Make sure all mind reading outputs are consumed
	if iterator.GetIndex() != len(block.Uncles()) {
		return nil, nil, 0, fmt.Errorf("unconsumed MR outputs, isReplay: %t", reuseMindReadingOutput)
	}
```

Comment (qizhou) [reply core/state_processor.go]: No, there are two cases
- All uncles are consumed, which is checked in line-127
- All outputs in one uncle are consumed, which I don't think we have such a check.

Comment (cyl19970726) [reply core/state_processor.go]: I already understood the TODO.
But I think the actual mindReadingOutput will be not equal to the preset mindreadingOutput  if all the mrOutputs are just partially consumed.

Comment (qizhou) [reply core/state_processor.go]: Looks like the current code just returns the `preset-mindreadingOutputs`?

File: core/vm/evm.go [hunk 2]
```diff
@@ -216,20 +217,31 @@ func (evm *EVM) IsMindReadingEnabled() bool {
...
-		evm.setCCCOutputs(versionedResult)
-		return true
+		return evm.setCCCOutputs(versionedResult)
 	}
-	return false
+	return nil
 }
 
-func (evm *EVM) GetCCCOutputs() []*CrossChainCallOutput {
+func (evm *EVM) GetCCCOutputs() ([]*CrossChainCallOutput, error) {
```
Comment (qizhou) [inline core/vm/evm.go:236]: I am not a big fan of this design - as a caller of the function, it should not require the knowledge of cCCOutputsIdx.

**Review by qizhou — COMMENTED** [review]
I think a simple change is to call getNextReplayableCCCOutput() outside (perhaps in state_processor), and check if it is not nil, then some CCCoutpus are not consumed.

**Comment by cyl19970726** [comment]
> I think a simple change is to call getNextReplayableCCCOutput() outside (perhaps in state_processor), and check if it is not nil, then some CCCoutpus are not consumed.

Good advice and update code to this mechanism at https://github.com/QuarkChain/go-ethereum/pull/121/commits/2c87ee7772ce0daeda8a5e5299caaac85c569756

File: core/state_processor.go [hunk 3]
```diff
@@ -114,6 +114,9 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 			} else {
 				// TODO: make sure all the mrOutput from a single transaction are consumed
 				// a non-validator fails to replay the MR data
+				if vmenv.GetNextReplayableCCCOutput() != nil {
+					return nil, nil, 0, fmt.Errorf("CCCOutpus is only partially consumed after completing a replay-mind-reading using the produced CCCOutpus", tx.Hash().Hex())
```
Comment (qizhou) [inline core/state_processor.go:118]: I guess you need to put "%s" in the error text?

**Review by qizhou — APPROVED** [review]

**Pull request merged**